### PR TITLE
Dev improve upload

### PR DESCRIPTION
**Pull request by bertmeeuws**
- Added file upload progress
- You can cancel an upload
- Added error messages for files bigger than 20mb or when something went wrong
- Added VueUse (@JonasWijne approved) to their useOnline(), an reactive online state. To let a person know if he/she is offline

Still one problem, if you upload a file and your connections breaks. State is gone, (don't know why yet).
But if you try to upload without a connection, you have an error. If you then have a connection and retry it works as it should.

image
image

### CODE REVIEW

File: src/store/chatStore.ts [hunk 1]
```diff
@@ -375,10 +376,21 @@ export const sendMessageObject = (chatId, message: Message<MessageBodyType>) =>
...
 };
 
+const delay = delayInms => {
+    return new Promise(resolve => {
+        setTimeout(() => {
+            resolve(2);
+        }, delayInms);
+    });
+};
+
```
Comment (JonasWijne) [inline src/store/chatStore.ts:386]: can be removed
```suggestion

```

File: src/store/chatStore.ts [hunk 2]
```diff
@@ -398,33 +412,143 @@ const sendFile = async (chatId, selectedFile, isBlob = false, isRecording = fals
...
+    if (e.message == 'Request failed with status code 413') {
+        //!TODO Upload limit
+        errorBody = 'ERROR: File exceeds 20MB limit!';
+        imageUploadQueue.value = imageUploadQueue.value.map(el => {
+            if (uuid === el.id) {
+                return { ...el, error_message: 'File exceeds 20MB limit!', error: true };
+            }
+            return el;
+        });
+    } else {
```
Comment (JonasWijne) [inline src/store/chatStore.ts:479]: :angry:

File: src/store/chatStore.ts [hunk 3]
```diff
@@ -398,33 +412,143 @@ const sendFile = async (chatId, selectedFile, isBlob = false, isRecording = fals
...
+
+        errorBody = 'ERROR: File failed to send!';
+        imageUploadQueue.value = imageUploadQueue.value.map(el => {
+            if (uuid === el.id) {
+                return { ...el, error_message: 'File failed to send', error: true, retry: true };
+            }
+            return el;
```
Comment (JonasWijne) [inline src/store/chatStore.ts:486]: duplicate code

File: src/store/chatStore.ts [hunk 4]
```diff
@@ -398,33 +412,143 @@ const sendFile = async (chatId, selectedFile, isBlob = false, isRecording = fals
...
+        });
+
+    }
+}
+
+export const retrySendFile = async (file) => {
+    if(!file) return;
```
Comment (JonasWijne) [inline src/store/chatStore.ts:493]: don't check if your input is valid, use typescript to your advantage and type the parameter, and make sure you never can call this function without the right data

File: src/store/chatStore.ts [hunk 5]
```diff
@@ -398,33 +412,143 @@ const sendFile = async (chatId, selectedFile, isBlob = false, isRecording = fals
...
+                return {
+                    ...el,
+                    error_message: '',
+                    loaded: 0,
+                    retry: false,
+                    error: false,
+                    cancelToken: source
+                }
+            }
+            return el
```
Comment (JonasWijne) [inline src/store/chatStore.ts:518]: duplicate code

File: src/store/chatStore.ts [hunk 6]
```diff
@@ -398,33 +412,143 @@ const sendFile = async (chatId, selectedFile, isBlob = false, isRecording = fals
...
+
+                    if (item.id === uuid) {
+                       return {
+                            ...item,
+                            loaded: progress,
+                            total: total,
+                            retry: false,
+                           error: false,
+                           error_message: ""
+                        };
```
Comment (JonasWijne) [inline src/store/chatStore.ts:538]: duplicate code

File: src/store/chatStore.ts [hunk 7]
```diff
@@ -398,33 +412,143 @@ const sendFile = async (chatId, selectedFile, isBlob = false, isRecording = fals
...
+
+                    }
+                    return item;
+                });
+            },
+        });
+    } catch (e) {
+        sendFileCatch(e, uuid)
```
Comment (JonasWijne) [inline src/store/chatStore.ts:546]: naming is not good, a function name should represent what its doing not from where it comes

**Review by JonasWijne — DISMISSED** [review]

**Review by ldruwel — APPROVED** [review]

**Pull request merged**